### PR TITLE
Added some modules for api & storage. Finished a-z auth and talk flow

### PR DESCRIPTION
**Pull request by Karagath**

### CODE REVIEW

File: extension/demo/src/bg/background.js [hunk 1]
```diff
@@ -23,19 +24,34 @@ function onToken (data, sendResponse) {
...
+}
+
+function init () {
+  chrome.storage.local.get(
+    [keys.auth],
+    result => {
+      if (!result[keys.auth]) {
+        openAuthTab()
+      } else {
+        setTimeout(updatePayload, timeToLive(result.payload.expiresAt))
```
Comment (leihog) [inline extension/demo/src/bg/background.js:45]: Shouldn't this be done after the token is received as well?

Comment (Karagath) [reply extension/demo/src/bg/background.js]: yes indeed

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+*.DS_Store
```
Comment (leihog) [inline .gitignore:1]: Although I don't mind that it is here, please also add that to your global gitignore. There is never a case when you want that included in a repo.

```
git config --global core.excludesfile ~/.gitignore
echo "*.DS_Store" > ~/.gitignore
```
🥃

File: extension/demo/src/bg/background.js [hunk 3]
```diff
@@ -23,44 +24,59 @@ function onPopupOpenApps () {
...
   const { subscription } = data
-  let payload = {}
-  payload[STORAGE_KEY]['auth']['subscription'] = subscription
-  await chrome.storage.local.set(payload)
-  chrome.storage.local.get([STORAGE_KEY], results => {
-    console.log(results)
-    return results
-  })
+  await storage.set({ subscription })
 }
```
Comment (gburning) [inline extension/demo/src/bg/background.js:60]: No need to use `await` here as there's nothing we need to wait for. If you want to return a promise I would instead remove the `async/await` and just return the `storage.set` function

File: extension/demo/src/bg/background.js [hunk 4]
```diff
@@ -23,44 +24,59 @@ function onPopupOpenApps () {
...
+  addListener('popup-trigger-auth', onPopupTriggerAuth)
+  addListener('popup-open-apps', onPopupOpenApps)
+
+  const accessToken = await storage.get(ACCESS_TOKEN)
+  if (accessToken) {
+    await updatePayload()
+  }
+}
+
+await init()
```
Comment (gburning) [inline extension/demo/src/bg/background.js:82]: This won't work. You can't use `await` outside an async function. Also there's not really any reason to wait here, there's nothing we need to block.

I would probably just make this a self invoking function instead:

```
;(async => {
  // ...
})()
```

File: extension/demo/src/modules/api.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
+'use strict'
+
+export function fetchPayload (accessToken) {
```
Comment (gburning) [inline extension/demo/src/modules/api.js:3]: Maybe a less generic name so it's clear that this is specifically for sub status.

Also you need to import `API_BASE`.

File: extension/demo/src/modules/constants.js [hunk 6]
```diff
@@ -2,5 +2,8 @@
 
 const env = process.env.TARGET_ENV
 
-export const STORAGE_KEY = 'auth'
+export const ACCESS_TOKEN = 'accessToken'
+export const SUBSCRIPTION = 'subscription'
+export const PAYLOAD = 'payload'
+export const TTL = 'expiresAt'
```
Comment (gburning) [inline extension/demo/src/modules/constants.js:8]: Maybe prefix them with something so it's clear what they're used for.

Though, frankly, it would probably be more useful to turn the whole storage module into a class/object with getters and setters for each (`getAccessToken`, `setAccessToken`, etc).

The latter might be better to do in a separate PR though.

**Review by gburning — CHANGES_REQUESTED** [review]
Added some comments.

You might want to have a look at and merge https://github.com/flattr/browser-extension-auth-flow-demo/pull/28 first. It implements https://github.com/mozilla/webextension-polyfill#using-the-promise-based-apis to make this whole extension business a lot easier to work with.

File: extension/demo/src/bg/background.js [hunk 7]
```diff
@@ -1,66 +1,85 @@
...
-addListener('popup-trigger-auth', onPopupTriggerAuth)
-addListener('popup-open-apps', onPopupOpenApps)
+;(async () => {
+  addListener('token', onToken)
+  addListener('subscription', onSubscription)
+  addListener('popup-check-auth', onPopupCheckAuth)
+  addListener('popup-trigger-auth', onPopupTriggerAuth)
+  addListener('popup-open-apps', onPopupOpenApps)
+
+  const accessToken = storage.get(ACCESS_TOKEN)
```
Comment (gburning) [inline extension/demo/src/bg/background.js:80]:
```suggestion
  const accessToken = await storage.get(ACCESS_TOKEN)
```

`storage.get` returns a promise, so you need to have an `await` here

File: extension/demo/src/modules/storage.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+      resolve(results[key])
+    })
+  })
+}
+
+export async function remove (key) {
+  await chrome.storage.local.remove(key)
+
+  return !chrome.runtime.lastError
+}
```
Comment (gburning) [inline extension/demo/src/modules/storage.js:23]: Should use `browser` instead of `chrome` for all of these. Also make sure to put `import browser from 'webextension-polyfill'` at the top.

File: extension/demo/src/modules/storage.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+export async function get (key) {
+  return new Promise(resolve => {
+    chrome.storage.local.get([key], results => {
+      console.log(results[key])
+      resolve(results[key])
+    })
+  })
+}
```
Comment (gburning) [inline extension/demo/src/modules/storage.js:17]: A Promise within an `async` function is generally bad practice. This can be rewritten (and simplified) as:

```
export function get (key) {
  return chrome.storage.local.get([key])
}
```

You're returning a promise either way (`async` always makes the return of a function a promise) so there's no need to complicate it.

File: extension/demo/src/modules/storage.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
+'use strict'
+
+export async function set (data) {
+  console.log(data)
+  await chrome.storage.local.set(data, console.log)
+
+  return !chrome.runtime.lastError
+}
```
Comment (gburning) [inline extension/demo/src/modules/storage.js:8]: Thanks to the [polyfill](https://github.com/mozilla/webextension-polyfill#using-the-promise-based-apis) we don't need to check for `lastError`. Instead, simply return the promise and if we need to handle rejections/failures we can simply nest the call to this function (`storage.set`) with a try/catch.

File: extension/demo/src/bg/background.js [hunk 11]
```diff
@@ -1,66 +1,85 @@
...
+    console.log(subscriptionStatus)
+    if (subscriptionStatus.hasOwnProperty(PAYLOAD)) {
+      storage.set(subscriptionStatus)
+    }
+  } else {
+    const ttl = await storage.get(TTL)
+    if (ttl) {
+      let timeout = timeToLive(ttl)
+      console.log(timeout)
+      setTimeout(updatePayload.bind(null, accessToken), timeout)
```
Comment (gburning) [inline extension/demo/src/bg/background.js:56]: Another, slightly cleaner, option:
```suggestion
      const callback = updatePayload(accessToken)
      setTimeout(callback, timeout)
```

**Review by gburning — CHANGES_REQUESTED** [review]

File: extension/demo/src/bg/background.js (hunk 7 above)
Comment (Karagath) [reply extension/demo/src/bg/background.js]: Ah, so await works then, I am confusion, the code is since changed to then() but would be cleaner like this

File: extension/demo/src/bg/background.js (hunk 11 above)
Comment (Karagath) [reply extension/demo/src/bg/background.js:56]: Argument type Promise<void> is not assignable to parameter type TimerHandler   
Type Promise<void> is not assignable to type string | Function    
Type Promise<void> is not assignable to type Function 

🤔

Comment (gburning) [reply extension/demo/src/bg/background.js]: My bad, it should of course return a wrapper function. I've updated the suggestion.

File: extension/demo/src/modules/storage.js (hunk 10 above)
Comment (Karagath) [reply extension/demo/src/modules/storage.js]: That seems less practical than returning a bool

File: extension/demo/src/modules/constants.js (hunk 6 above)
Comment (Karagath) [reply extension/demo/src/modules/constants.js]: https://github.com/flattr/browser-extension-auth-flow-demo/issues/30

**Comment by Karagath** [comment]
I think GitHub is misunderstanding a commit comment about the merge as listing a present conflict 🙄

File: extension/demo/src/bg/background.js [hunk 12]
```diff
@@ -3,64 +3,80 @@
...
-    isAuthenticated
+  const { accessToken } = data
+  const isAuthenticated = await storage.set(data)
+  updatePayload(accessToken)
+
+  return { isAuthenticated }
+}
+
+async function requestPayload () {
+  let response = {}
```
Comment (gburning) [inline extension/demo/src/bg/background.js:36]:
```suggestion
  const response = {}
```

No reason to use `let` here.

File: extension/demo/src/modules/storage.js [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+
+  return !browser.runtime.lastError
+}
+
+export function get (key) {
+  return new Promise(async resolve => {
+    const results = await browser.storage.local.get([key])
+    resolve(results[key])
+  })
+}
```
Comment (gburning) [inline extension/demo/src/modules/storage.js:16]:
```suggestion
export async function get (key) {
  const results = await browser.storage.local.get([key])
  return results[key]
}
```

File: extension/demo/src/bg/background.js [hunk 14]
```diff
@@ -3,64 +3,80 @@
...
+      setTimeout(callback, timeToLive(ttl))
+    }
   }
 }
 
-function onSubscription (data) {
-  updateStorage(data)
+async function onSubscription (data) {
+  storage.set(data)
 }
```
Comment (gburning) [inline extension/demo/src/bg/background.js:64]:
```suggestion
function onSubscription (data) {
  storage.set(data)
}
```

The function is defined as async but we're not waiting on anything.

**Review by leihog — APPROVED** [review]

File: extension/demo/src/modules/storage.js (hunk 10 above)
Comment (gburning) [reply extension/demo/src/modules/storage.js]: It isn't. The above just magics away any error messages which we might want to do some times but definitely not always.

Most places that consume this function would simply do `const results = await storage.set(...)` without any fallback. If we want to make sure that we recover from any potential errors in production we should probably do that globally or at least via some sort helper/decoration. As of now there's not really any reason to though.

The only place where we need/want the error to be handled right now is in the response to setting the access token. It's better to handle it there.

File: extension/demo/src/modules/storage.js (hunk 13 above)
Comment (Karagath) [reply extension/demo/src/modules/storage.js:16]: This looks like the code that we had which did not work but we can try again

Comment (gburning) [reply extension/demo/src/modules/storage.js]: Sounds strange. What wasn't working?

**Review by gburning — APPROVED** [review]
